In [3]:
import warnings
warnings.filterwarnings('ignore')


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
!pip install mlxtend

In [1]:
import pandas as pd
from mlxtend.frequent_patterns import apriori,association_rules

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [4]:
data = {
    'Bread':[1,1,1,0,1],
    'Butter':[1,1,0,1,1],
    'Milk':[1,0,1,0,0],
    'Jam':[0,0,0,1,1]
}
df = pd.DataFrame(data)
df

Bread  Butter  Milk  Jam
0      1       1     1    0
1      1       1     0    0
2      1       0     1    0
3      0       1     0    1
4      1       1     0    1

In [6]:
frequent_itemset = apriori(df,min_support=0.4,use_colnames=True)
frequent_itemset

support         itemsets
0      0.8          (Bread)
1      0.8         (Butter)
2      0.4           (Milk)
3      0.4            (Jam)
4      0.6  (Butter, Bread)
5      0.4    (Milk, Bread)
6      0.4    (Butter, Jam)

In [ ]:
# use_colnames True means the output will show item names instead of numbers

In [14]:
rules = association_rules(frequent_itemset,metric='confidence',min_threshold=0.6)
rules[['antecedents','consequents','support','confidence','lift']]

antecedents consequents  support  confidence    lift
0    (Butter)     (Bread)      0.6        0.75  0.9375
1     (Bread)    (Butter)      0.6        0.75  0.9375
2      (Milk)     (Bread)      0.4        1.00  1.2500
3       (Jam)    (Butter)      0.4        1.00  1.2500

In [ ]:
# confidence - to focus on how often the rule is true
# antecedents - the if part of rule
# consequents - the then part of the rule
# support - how often the itmes in the rule appear together
# lift -- how much more likely the items are bought together compared to seperately(lift>1 means a strong pattern)


Bread and Butter are strongly linked:

Rule: "If Bread, then Butter" has 60% support (they appear together in 3 lists) and 75% confidence (75% of people who buy Bread also buy Butter).

The reverse rule ("If Butter, then Bread") is also true with 75% confidence.

Insight: Customers who buy Bread are very likely to buy Butter, and vice versa. A store could bundle these items or place them close together.

Note: The lift (0.9375) is slightly less than 1, suggesting the connection is strong but not much stronger than if Bread and Butter were bought independently (since both are very common). It means Are people really buying two items together because they are connected, or is it just by chance?"

Lift > 1 → People are buying them together more than expected → There’s a real connection.

Lift = 1 → Buying together is just coincidence.

Lift < 1 → Buying together happens less than expected → Probably no real connection.